pkg/xds/experimental/client/watcher.go

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package client
+
+import (
+	"log/slog"
+
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// watchers is a helper structure for Client focused on [un]registering callbacks on resource changes.
+type watchers struct {
+	log *slog.Logger
+
+	src xds.ObservableResourceSource
+
+	// mux protects all fields below.
+	mux lock.Mutex
+	// listeners maps ID to an instance of a listener.
+	listeners map[uint64]*listener
+	lastID    uint64
+}
+
+func newWatchers(log *slog.Logger, src xds.ObservableResourceSource) *watchers {
+	return &watchers{
+		log:       log,
+		src:       src,
+		listeners: make(map[uint64]*listener),
+	}
+}
+
+// WatcherCallback will be called when a new version of a resource it was registered on appears.
+type WatcherCallback func(res *xds.VersionedResources)
+
+// Add registers a callback for a specified typeUrl.
+// Returned ID can be used to later unregister the callback.
```

Comment:
if the id is just an opaque handle, I suggest typing it to make this obvious:

```
type watcherHandle uint64
```

Reply:
Done.